lib/extract_field/FieldExtractor.js

```diff
@@ -0,0 +1,42 @@
+const { sha256, accessNestedObjectValueByStringPath } = require("../pxutil");
+
+class FieldExtractor {
+    static SENT_THROUGH = { BODY: "body", QUERY_PARAMS: "query-param", HEADERS: "header" };
+    static CONTENT_TYPE = { JSON: "json", XML: "xml", FORM_DATA: "form-data"};
+    static CONTAINER = { "body": "body", "query-param": "query", "header": "headers"};
+    static ENCODING = { CLEAR_TEXT: "clear-text", URL_ENCODE: "url-encode", BASE64: "base64", CUSTOM: "custom" };
+
+    constructor(sentThrough, contentType, encoding, fieldsToExtract) {
+        this.containerName = this._getContainerName(sentThrough);
+        this.contentType = contentType;
+        this.encoding = encoding;
+        this.fieldsToExtract = fieldsToExtract;
+    }
+
+    ExtractFields(request) {
+        const fields = {};
+        const container = request[this.containerName];
+        if (!container) {
+            return fields
+        }
+        this.fieldsToExtract.forEach((desiredField) => {
+            const value = accessNestedObjectValueByStringPath(desiredField.bodyFieldName, container);
+            if (!value || typeof value !== "string") {
+                return;
```

Comment:
why not return empty object as above? I think it is better to have single failure output

Reply:
This is a callback, so if we don't have a value, we don't need to do anything for that particular desired field. No need to provide any return value since the callback function `(desiredField) => { ... }` always returns undefined.